Add Github Action to build and publish Docker Image to ghcr.io

itoumlilt opened the pull request:
GA workflow that builds Strimzi image and publish it to ghcr.io/criteo-fork/strimzi-kafka-operator with tag based on trigger:
- merge to criteo-0.42.x -> latest tag
- git tag release 0.42.* -> stable tag + "0.42.*" tag
- opened or updated pull request -> dev tag with pull request ID

itoumlilt committed:
Add Github Action to build and publish Focker Image to ghcr.io

itoumlilt committed:
Fix typo

kooli89 commented on .github/workflows/build-and-push-image.yml:
```diff
@@ -0,0 +1,81 @@
+            ${{ runner.os }}-maven-
+
+      - name: Build with Maven
+        run: mvn -B -V -e -Dfindbugs.skip -Dcheckstyle.skip -Dpmd.skip=true -Denforcer.skip -Dmaven.javadoc.skip -DskipTests -Dmaven.test.skip.exec -Dlicense.skip=true -Drat.skip=true clean install
```
I suggest to either two options:
1- Have a dedicated action to build and test the project that precedes the image push.
2- Build and test the project here. 

Building and testing the project will act +- the same as the pre-submit we have in gerrit.

kooli89 commented on .github/workflows/build-and-push-image.yml:
```diff
@@ -0,0 +1,81 @@
+on:
+  push:
+    branches:
+      - criteo-0.42.x
```
Does it mean we don't trigger the action when we push to other branches?
For me this action should be triggered for push in other branches to be able to iterate and test during the dev phase.

itoumlilt replied:
No, it triggers on any pull request as a CI verification step (you can see it failing for the current Pull Request that is from `build-and-push-image-workflow` branch).
But also from commit/tag push to `criteo-0.42.x`, and for this the release step acts differently -> tagging the image as "latest" for merge or "stable" for git tag push.

itoumlilt committed:
Fix ref checkout

itoumlilt committed:
Fix tagging

itoumlilt committed:
Separate CI & CD workflows

itoumlilt committed:
CI-workflow: remove pull_request trigger
Already covered by push trigger.

itoumlilt committed:
CI/CD: Fix context for docker push action

itoumlilt committed:
CI/CD: Use Makefile.docker instead of raw docker/build-push-action@v3

itoumlilt committed:
CI: Only one tag is necessary as no image push

itoumlilt committed:
CI: Fix docker tag ref

itoumlilt committed:
CI: Add build env to Workflow

itoumlilt committed:
CI: List available docker images

itoumlilt committed:
CI: Fix usage of `make docker_tag`

itoumlilt committed:
CI: Fix usage of `make docker_tag`

itoumlilt committed:
CD: Add extra Docker tag for latest/stable cases

# Spec:
Based on triggering event, determine the image tag to use
- For PR merge: use the commit SHA + 'latest'
- For git tag push: use the tag name + 'stable'
- For branch push: use the commit SHA + 'latest'
- For manual trigger: use the provided commit SHA
